src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseExplicitTypingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,91 @@
+            title: s_Title,
+            messageFormat: s_Message,
+            category: DiagnosticCategory.Style,
+            defaultSeverity: DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        public CSharpUseExplicitTypingDiagnosticAnalyzer() : base(s_descriptorUseImplicitTyping)
+        {
+
+        }
+
+        protected override bool IsStylePreferred(SyntaxNode declarationStatement, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
+        {
+            var stylePreferences = GetCurrentTypingStylePreferences(optionSet);
+
+            var isTypeApparent = IsTypeApparentInDeclaration(declarationStatement, semanticModel, stylePreferences, cancellationToken);
+            var isIntrinsicType = IsIntrinsicType(declarationStatement);
+
+            return stylePreferences.HasFlag(TypingStyles.NoVarForIntrinsic) && isIntrinsicType
+                || stylePreferences.HasFlag(TypingStyles.NoVarWhereApparent) && isTypeApparent
+                || stylePreferences.HasFlag(TypingStyles.NoVarWherePossible) && !(isIntrinsicType || isTypeApparent);
+        }
+
+        protected override bool AnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan)
+        {
+            issueSpan = default(TextSpan);
+
+            // If it is currently not var, explicit typing exists, return. 
+            // this also takes care of cases where var is mapped to a named type via an alias or a class declaration.
+            if (!typeName.IsTypeInferred(semanticModel))
+            {
+                return false;
+            }
+
+            if (typeName.Parent.IsKind(SyntaxKind.VariableDeclaration) &&
+                typeName.Parent.Parent.IsKind(SyntaxKind.LocalDeclarationStatement, SyntaxKind.ForStatement, SyntaxKind.UsingStatement))
+            {
+                // check assignment for variable declarations.
+                var variable = ((VariableDeclarationSyntax)typeName.Parent).Variables.First();
+                if (!AnalyzeAssignment(variable.Identifier, typeName, variable.Initializer, semanticModel, optionSet, cancellationToken))
+                {
+                    return false;
+                }
+            }
+
+            issueSpan = typeName.Span;
+            return true;
+        }
+
+        protected override bool AnalyzeAssignment(SyntaxToken identifier, TypeSyntax typeName, EqualsValueClauseSyntax initializer, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
+        {
+            // is or contains an anonymous type
+            // cases :
+            //        var anon = new { Num = 1 };
+            //        var enumerableOfAnons = from prod in products select new { prod.Color, prod.Price };
+            var declaredType = semanticModel.GetTypeInfo(typeName, cancellationToken).Type;
```

Comment:
Does calling `GetTypeInfo` on the type name versus the initializer expression return different information? If not, you could probably call it on just one and check that for being an error/containing anonymous types.